{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation file\n",
    "\n",
    "The lncRNA transcripts were given as a .fasta file.\n",
    "\n",
    "Firstly, sequences in fasta files were renamed, because the names were too long. The renamed fasta was aligned to the genome, turned to bed and gtf produced out of the bed file was used. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"merged_lncRNAs_transcripts_filtered_nonredundant.out_30_1_plus\", \"r\") as infa:\n",
    "    with open(\"merged_lncRNAs_transcripts_filtered_nonredundant_new.fasta\", \"w\") as outfa:\n",
    "        with open(\"merged_lncRNAs_transcripts_filtered_nonredundant_tocutfasta.csv\", \"w\") as pot:\n",
    "            i = 0\n",
    "            for line in infa:\n",
    "                if line.startswith(\">\"):\n",
    "                    i += 1\n",
    "                    outfa.write(\">lnc_tr\")\n",
    "                    outfa.write(str(i) + \"\\n\")\n",
    "                    pot.write(line.strip() + \";\" + \">lnc_tr\" + str(i) + \"\\n\") # table of new and old names matching \n",
    "                else:\n",
    "                    outfa.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Flair needs reads, gtf annotation and a reference. Gtf annotation was produced from fasta file of lncRNAs with **fqtobed.sh** script. Then bed12ToGTF from https://github.com/gold-lab/bed12ToGTF was used to produce gtf. The resulted file lacked gene_id, so the following commands were used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed 's/transcript/gene/' lnctogtf.gtf > lnctogtf_genes.gtf\n",
    "!cut -f 9 lnctogtf_genes.gtf | sed 's/;/\";/g' | sed 's/number /number \"/' > lnctogtf_genes_column\n",
    "!sed 's/;/; transcript_id/' lnctogtf_genes_column > lnctogtf_genes_column2\n",
    "!awk '{print $1, $2, $3, $2, $4, $5, $6}' lnctogtf_genes_column2 > lnctogtf_genes_column_fin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv('lnctogtf_genes.gtf', sep='\\t', header = None)\n",
    "fin = pd.read_csv('lnctogtf_genes_column_fin', sep='\\t', header = None)\n",
    "gtf.iloc[:,8] = fin.iloc[:,0]\n",
    "gtf.to_csv(\"lnc_annot_1.gtf\", sep = \"\\t\", header=False, index=False)\n",
    "!sed 's/\"\"/\"/g' lnc_annot_1.gtf | sed 's/\"gene/gene/' | sed 's/;   \"/;/' | sed 's/; \"/;/' > lnc_annot.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reads\n",
    "\n",
    "Basecalling was performed using Guppy.\n",
    "\n",
    "#### guppy_basecaller --input_path fast5/g3 --num_callers 22 --save_path /media/fastq/moss/g3 --flowcell FLO-MIN106 --kit SQK-RNA002 --qscore_filtering -r\n",
    "\n",
    "\n",
    ".bed12 files were produced with **fqtobed.sh** script.\n",
    "Then all .bed files were merged into 1 file in the following manner:\n",
    "\n",
    "#### cat p?.bed >> guppy4all.bed\n",
    "\n",
    "Then the file was sorted:\n",
    "\n",
    "#### bedtools sort -i guppy4all.bed > guppy4all_sort.bed\n",
    "\n",
    "All .fastq reads were copied to a single file for each replication and into a file with all reads of all replications.\n",
    "\n",
    "#### cat guppy4/g2/*.fastq >> g2.fastq\n",
    "#### cat guppy4/g2/*.fastq >> all_seqs.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flair\n",
    "\n",
    "Flair of the latest version from https://github.com/BrooksLabUCSC/flair was used. However, an older version was used for **flair correct** command. The pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sudo docker exec 537efdb1ac27 python3 /usr/local/flair/flair.py correct -t 18 -o lnc -q guppy4all_sort.bed -g moss_ref.fasta -f lnc_annot.gtf -c sizes.genome\n",
    "!python flair.py collapse -f lnc_annot.gtf --generate_map -o lnc -q lnc_corrected.bed -g moss_ref.fasta -r all_seqs.fastq -s 1 -t 18\n",
    "!python ~/anna/flair/flair.py quantify -t 18 -o lnc_newgtf2 -r reads_manifest.tsv -i lnc.isoforms.fa\n",
    "!grep lnc lnc_newgtf2 | sed 's/.*_lnc/lnc/' > lnc_newgtf2_codes.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr14435</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr4765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr15475</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr8473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr12486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0    1    2    3    4     5    6     7\n",
       "0  lnc_tr14435  0.0  0.0  1.0  0.0   0.0  0.0   0.0\n",
       "1   lnc_tr4765  0.0  0.0  0.0  0.0   1.0  0.0   1.0\n",
       "2  lnc_tr15475  0.0  0.0  7.0  0.0   0.0  0.0   0.0\n",
       "3   lnc_tr8473  0.0  0.0  0.0  0.0   1.0  0.0   1.0\n",
       "4  lnc_tr12486  0.0  0.0  0.0  0.0  11.0  1.0  13.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flair = pd.read_csv('lnc_newgtf2_codes.tsv', sep='\\t', header = None)\n",
    "flair.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;CNT2063571</td>\n",
       "      <td>lnc_tr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;CNT2063572</td>\n",
       "      <td>lnc_tr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;CNT2063573</td>\n",
       "      <td>lnc_tr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;CNT2063574</td>\n",
       "      <td>lnc_tr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;CNT2063575</td>\n",
       "      <td>lnc_tr5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0        1\n",
       "0  >CNT2063571  lnc_tr1\n",
       "1  >CNT2063572  lnc_tr2\n",
       "2  >CNT2063573  lnc_tr3\n",
       "3  >CNT2063574  lnc_tr4\n",
       "4  >CNT2063575  lnc_tr5"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now we can return full transcript names\n",
    "names = pd.read_csv('lnc_tr_names_tomynames.csv', sep=';', header=None)\n",
    "names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "flair = flair.rename(columns = {0: \"ids\",1: \"g1_gam\",2: \"g2_gam\",3: \"g3_gam\",4: \"g4_gam\", 5: \"p1_prot\", 6: \"p2_prot\",7: \"p3_prot\"})\n",
    "names = names.rename(columns={1: \"ids\", 0: \"names\"})\n",
    "flair = flair.merge(names, how=\"inner\", on=\"ids\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>g1_gam</th>\n",
       "      <th>g2_gam</th>\n",
       "      <th>g3_gam</th>\n",
       "      <th>g4_gam</th>\n",
       "      <th>p1_prot</th>\n",
       "      <th>p2_prot</th>\n",
       "      <th>p3_prot</th>\n",
       "      <th>names</th>\n",
       "      <th>g1_rpm</th>\n",
       "      <th>g2_rpm</th>\n",
       "      <th>g3_rpm</th>\n",
       "      <th>g4_rpm</th>\n",
       "      <th>p1_rpm</th>\n",
       "      <th>p2_rpm</th>\n",
       "      <th>p3_rpm</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr14435</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.298701</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>flair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr14435</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.05499</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>flair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr4765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_40640V3.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>flair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr15475</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;XR_002970609.1 rna-XR_002970609.1_NCBI_ncRNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>9.090909</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>flair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr8473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c6_15880V3.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>flair</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ids  g1_gam  g2_gam  g3_gam  g4_gam  p1_prot  p2_prot  p3_prot  \\\n",
       "0  lnc_tr14435     0.0     0.0     1.0     0.0      0.0      0.0      0.0   \n",
       "1  lnc_tr14435     0.0     3.0     0.0     0.0      1.0      0.0      0.0   \n",
       "2   lnc_tr4765     0.0     0.0     0.0     0.0      1.0      0.0      1.0   \n",
       "3  lnc_tr15475     0.0     0.0     7.0     0.0      0.0      0.0      0.0   \n",
       "4   lnc_tr8473     0.0     0.0     0.0     0.0      1.0      0.0      1.0   \n",
       "\n",
       "                                           names  g1_rpm   g2_rpm    g3_rpm  \\\n",
       "0  >XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA     0.0  0.00000  1.298701   \n",
       "1  >XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA     0.0  3.05499  0.000000   \n",
       "2                   >lcl|Ppatens_Pp3c1_40640V3.1     0.0  0.00000  0.000000   \n",
       "3  >XR_002970609.1 rna-XR_002970609.1_NCBI_ncRNA     0.0  0.00000  9.090909   \n",
       "4                   >lcl|Ppatens_Pp3c6_15880V3.1     0.0  0.00000  0.000000   \n",
       "\n",
       "   g4_rpm    p1_rpm  p2_rpm    p3_rpm source  \n",
       "0     0.0  0.000000     0.0  0.000000  flair  \n",
       "1     0.0  0.626174     0.0  0.000000  flair  \n",
       "2     0.0  0.626174     0.0  0.552181  flair  \n",
       "3     0.0  0.000000     0.0  0.000000  flair  \n",
       "4     0.0  0.626174     0.0  0.552181  flair  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flair[\"source\"] = \"flair\"\n",
    "flair.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The number of reads is from samtools flagstat\n",
    "flair[\"g1_rpm\"] = flair.g1_gam/0.393\n",
    "flair[\"g2_rpm\"] = flair.g2_gam/0.982\n",
    "flair[\"g3_rpm\"] = flair.g3_gam/0.77\n",
    "flair[\"g4_rpm\"] = flair.g4_gam/0.88\n",
    "flair[\"p1_rpm\"] = flair.p1_prot/1.597\n",
    "flair[\"p2_rpm\"] = flair.p2_prot/0.5656\n",
    "flair[\"p3_rpm\"] = flair.p3_prot/1.811"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## featureCounts\n",
    "\n",
    "Flair cannot correctly identify isoforms with no introns. For monoexon features identification we used **featureCounts**. Monoexon features were extracted like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"lnc_annot.gtf\", \"r\") as gtf:\n",
    "    with open(\"lnc_annot_1ex.gtf\", \"w\") as mono:\n",
    "        for line in gtf.readlines():\n",
    "            line = line.split(\"\\t\")\n",
    "            if line[2] == \"gene\":\n",
    "                gene = line\n",
    "            else:\n",
    "                if line[3] == gene[3] and line[4] == gene[4]: # first exon is the same as the whole transcript\n",
    "                    mono.write(\"\\t\".join(gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "featureCounts takes annotation in a .saf format. We used a threshold of 75% for the feature to be covered by read."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $10, $3, \"\\t\" $1, \"\\t\", $4, \"\\t\", $5, \"\\t\", $7}' lnc_annot.gtf > lnc_annotm.saf\n",
    "!sed 's/ //g' lnc_annotm.saf > lnc_annot.saf\n",
    "!./featureCounts -a lnc_annot_1e.saf -F SAF -s 1 -L -O -o lnc_monoexon75 --fracOverlapFeature 0.75 --fracOverlap 0.75 g1.bam g2.bam g3.bam g4.bam p1.bam p2.bam p3.bam &\n",
    "!sed 's/;gene//' lnc_monoexon75 > lnc_monoexon75.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>/home/andrey/anna/g1.bam</th>\n",
       "      <th>/home/andrey/anna/g2.bam</th>\n",
       "      <th>/home/andrey/anna/g3.bam</th>\n",
       "      <th>/home/andrey/anna/g4.bam</th>\n",
       "      <th>/home/andrey/anna/p1.bam</th>\n",
       "      <th>/home/andrey/anna/p2.bam</th>\n",
       "      <th>/home/andrey/anna/p3.bam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr17504</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>185390</td>\n",
       "      <td>186013</td>\n",
       "      <td>+</td>\n",
       "      <td>624</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr4790</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>237391</td>\n",
       "      <td>237672</td>\n",
       "      <td>-</td>\n",
       "      <td>282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr4852</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627880</td>\n",
       "      <td>628299</td>\n",
       "      <td>+</td>\n",
       "      <td>420</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr14183</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627880</td>\n",
       "      <td>628299</td>\n",
       "      <td>+</td>\n",
       "      <td>420</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr14184</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627995</td>\n",
       "      <td>628180</td>\n",
       "      <td>+</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Geneid    Chr   Start     End Strand  Length  \\\n",
       "0  lnc_tr17504  Chr01  185390  186013      +     624   \n",
       "1   lnc_tr4790  Chr01  237391  237672      -     282   \n",
       "2   lnc_tr4852  Chr01  627880  628299      +     420   \n",
       "3  lnc_tr14183  Chr01  627880  628299      +     420   \n",
       "4  lnc_tr14184  Chr01  627995  628180      +     186   \n",
       "\n",
       "   /home/andrey/anna/g1.bam  /home/andrey/anna/g2.bam  \\\n",
       "0                         0                         0   \n",
       "1                         0                         0   \n",
       "2                         0                         0   \n",
       "3                         0                         0   \n",
       "4                         0                         0   \n",
       "\n",
       "   /home/andrey/anna/g3.bam  /home/andrey/anna/g4.bam  \\\n",
       "0                         0                         0   \n",
       "1                         0                         0   \n",
       "2                         0                         0   \n",
       "3                         0                         0   \n",
       "4                         0                         0   \n",
       "\n",
       "   /home/andrey/anna/p1.bam  /home/andrey/anna/p2.bam  \\\n",
       "0                         0                         0   \n",
       "1                         0                         0   \n",
       "2                         0                         0   \n",
       "3                         0                         0   \n",
       "4                         0                         0   \n",
       "\n",
       "   /home/andrey/anna/p3.bam  \n",
       "0                         0  \n",
       "1                         0  \n",
       "2                         0  \n",
       "3                         0  \n",
       "4                         0  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc = pd.read_csv('lnc_monoexon75.tsv', sep='\\t')\n",
    "fc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1347, 14)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc[\"summa\"] = fc.iloc[:,6] + fc.iloc[:,7] + fc.iloc[:,8] + fc.iloc[:,9] + fc.iloc[:,10] + fc.iloc[:,11] + fc.iloc[:,12]\n",
    "# only identified isoforms\n",
    "fc = fc[fc.summa > 0]\n",
    "fc.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc[\"g1_rpm\"] = fc.iloc[:,6]/0.393\n",
    "fc[\"g2_rpm\"] = fc.iloc[:,7]/0.982\n",
    "fc[\"g3_rpm\"] = fc.iloc[:,8]/0.77\n",
    "fc[\"g4_rpm\"] = fc.iloc[:,9]/0.88\n",
    "fc[\"p1_rpm\"] = fc.iloc[:,10]/1.597\n",
    "fc[\"p2_rpm\"] = fc.iloc[:,11]/0.5656\n",
    "fc[\"p3_rpm\"] = fc.iloc[:,12]/1.811\n",
    "fc[\"source\"] = \"featureCounts\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>/home/andrey/anna/g1.bam</th>\n",
       "      <th>/home/andrey/anna/g2.bam</th>\n",
       "      <th>/home/andrey/anna/g3.bam</th>\n",
       "      <th>/home/andrey/anna/g4.bam</th>\n",
       "      <th>...</th>\n",
       "      <th>summa</th>\n",
       "      <th>g1_rpm</th>\n",
       "      <th>g2_rpm</th>\n",
       "      <th>g3_rpm</th>\n",
       "      <th>g4_rpm</th>\n",
       "      <th>p1_rpm</th>\n",
       "      <th>p2_rpm</th>\n",
       "      <th>p3_rpm</th>\n",
       "      <th>names_x</th>\n",
       "      <th>names_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr9777</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>1852341</td>\n",
       "      <td>1854241</td>\n",
       "      <td>-</td>\n",
       "      <td>1901</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>2.544529</td>\n",
       "      <td>1.018330</td>\n",
       "      <td>11.688312</td>\n",
       "      <td>7.954545</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;PACX32971569 gene_name=Pp3c1_2520 ref_gene_id...</td>\n",
       "      <td>&gt;PACX32971569 gene_name=Pp3c1_2520 ref_gene_id...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr4686</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>2764720</td>\n",
       "      <td>2765825</td>\n",
       "      <td>-</td>\n",
       "      <td>1106</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.136364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_3540V3.1</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_3540V3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr14192</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>2764720</td>\n",
       "      <td>2765825</td>\n",
       "      <td>-</td>\n",
       "      <td>1106</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.136364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;PACX32968253 gene_name=Pp3c1_3540 ref_gene_id...</td>\n",
       "      <td>&gt;PACX32968253 gene_name=Pp3c1_3540 ref_gene_id...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr4772</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>3100347</td>\n",
       "      <td>3101036</td>\n",
       "      <td>+</td>\n",
       "      <td>690</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>7.633588</td>\n",
       "      <td>10.183299</td>\n",
       "      <td>12.987013</td>\n",
       "      <td>11.363636</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_4120V3.1</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_4120V3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr763</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>3292977</td>\n",
       "      <td>3294719</td>\n",
       "      <td>-</td>\n",
       "      <td>1743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.136364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;CNT2064367</td>\n",
       "      <td>&gt;CNT2064367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ids    Chr    Start      End Strand  Length  \\\n",
       "0   lnc_tr9777  Chr01  1852341  1854241      -    1901   \n",
       "1   lnc_tr4686  Chr01  2764720  2765825      -    1106   \n",
       "2  lnc_tr14192  Chr01  2764720  2765825      -    1106   \n",
       "3   lnc_tr4772  Chr01  3100347  3101036      +     690   \n",
       "4    lnc_tr763  Chr01  3292977  3294719      -    1743   \n",
       "\n",
       "   /home/andrey/anna/g1.bam  /home/andrey/anna/g2.bam  \\\n",
       "0                         1                         1   \n",
       "1                         0                         0   \n",
       "2                         0                         0   \n",
       "3                         3                        10   \n",
       "4                         0                         0   \n",
       "\n",
       "   /home/andrey/anna/g3.bam  /home/andrey/anna/g4.bam  ...  summa    g1_rpm  \\\n",
       "0                         9                         7  ...     19  2.544529   \n",
       "1                         0                         1  ...      1  0.000000   \n",
       "2                         0                         1  ...      1  0.000000   \n",
       "3                        10                        10  ...     33  7.633588   \n",
       "4                         0                         1  ...      1  0.000000   \n",
       "\n",
       "      g2_rpm     g3_rpm     g4_rpm    p1_rpm  p2_rpm  p3_rpm  \\\n",
       "0   1.018330  11.688312   7.954545  0.626174     0.0     0.0   \n",
       "1   0.000000   0.000000   1.136364  0.000000     0.0     0.0   \n",
       "2   0.000000   0.000000   1.136364  0.000000     0.0     0.0   \n",
       "3  10.183299  12.987013  11.363636  0.000000     0.0     0.0   \n",
       "4   0.000000   0.000000   1.136364  0.000000     0.0     0.0   \n",
       "\n",
       "                                             names_x  \\\n",
       "0  >PACX32971569 gene_name=Pp3c1_2520 ref_gene_id...   \n",
       "1                        >lcl|Ppatens_Pp3c1_3540V3.1   \n",
       "2  >PACX32968253 gene_name=Pp3c1_3540 ref_gene_id...   \n",
       "3                        >lcl|Ppatens_Pp3c1_4120V3.1   \n",
       "4                                        >CNT2064367   \n",
       "\n",
       "                                             names_y  \n",
       "0  >PACX32971569 gene_name=Pp3c1_2520 ref_gene_id...  \n",
       "1                        >lcl|Ppatens_Pp3c1_3540V3.1  \n",
       "2  >PACX32968253 gene_name=Pp3c1_3540 ref_gene_id...  \n",
       "3                        >lcl|Ppatens_Pp3c1_4120V3.1  \n",
       "4                                        >CNT2064367  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc = fc.rename(columns = {\"Geneid\": \"ids\"})\n",
    "fc = fc.merge(names, how=\"inner\", on=\"ids\")\n",
    "fc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "# Now we can concatenate the two tables\n",
    "flair_cut = flair.drop(columns = [\"g1_gam\", \"g2_gam\", \"g3_gam\",\"g4_gam\", \"p1_prot\", \"p2_prot\",\"p3_prot\", \"ids\"])\n",
    "fc_cut = fc.iloc[:, 14:]\n",
    "outp = pd.concat([flair_cut, fc_cut])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Antisence\n",
    "\n",
    "Reads from the opposite strand were taken in a similar way with featureCounts.\n",
    "\n",
    "#### ./featureCounts -a lnc_annot.saf -F SAF -s 2 -L -O -o lnc_anti_newdata g1.bam g2.bam g3.bam g4.bam p1.bam p2.bam p3.bam &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr17504</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>185390</td>\n",
       "      <td>186013</td>\n",
       "      <td>+</td>\n",
       "      <td>624</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr4769</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>201028</td>\n",
       "      <td>201445</td>\n",
       "      <td>+</td>\n",
       "      <td>418</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr4790</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>237391</td>\n",
       "      <td>237672</td>\n",
       "      <td>-</td>\n",
       "      <td>282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr4852</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627880</td>\n",
       "      <td>628299</td>\n",
       "      <td>+</td>\n",
       "      <td>420</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr14183</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627880</td>\n",
       "      <td>628299</td>\n",
       "      <td>+</td>\n",
       "      <td>420</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ids      1       2       3  4    5  6  7  8  9  10  11  12\n",
       "0  lnc_tr17504  Chr01  185390  186013  +  624  0  0  0  0   0   0   1\n",
       "1   lnc_tr4769  Chr01  201028  201445  +  418  0  0  0  0   0   0   0\n",
       "2   lnc_tr4790  Chr01  237391  237672  -  282  0  0  0  0   0   0   0\n",
       "3   lnc_tr4852  Chr01  627880  628299  +  420  0  0  0  0   0   0   0\n",
       "4  lnc_tr14183  Chr01  627880  628299  +  420  0  0  0  0   0   0   0"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anti = pd.read_csv(\"lnc_anti_newdata.tsv\", sep=\"\\t\", header=None)\n",
    "anti = anti.rename(columns={0: \"ids\"})\n",
    "anti.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "anti =  anti.merge(names, how=\"inner\", on=\"ids\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>names</th>\n",
       "      <th>g1_anti_rpm</th>\n",
       "      <th>g2_anti_rpm</th>\n",
       "      <th>g3_anti_rpm</th>\n",
       "      <th>g4_anti_rpm</th>\n",
       "      <th>p1_anti_rpm</th>\n",
       "      <th>p2_anti_rpm</th>\n",
       "      <th>p3_anti_rpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr17504</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>185390</td>\n",
       "      <td>186013</td>\n",
       "      <td>+</td>\n",
       "      <td>624</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>&gt;Pp3c1_350V3::Chr01:185389-186013(+)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr4769</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>201028</td>\n",
       "      <td>201445</td>\n",
       "      <td>+</td>\n",
       "      <td>418</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_410V3.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr4790</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>237391</td>\n",
       "      <td>237672</td>\n",
       "      <td>-</td>\n",
       "      <td>282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_500V3.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr4852</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627880</td>\n",
       "      <td>628299</td>\n",
       "      <td>+</td>\n",
       "      <td>420</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_870V3.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr14183</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>627880</td>\n",
       "      <td>628299</td>\n",
       "      <td>+</td>\n",
       "      <td>420</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&gt;PACX32968721 gene_name=Pp3c1_870 ref_gene_id=...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ids      1       2       3  4    5  6  7  8  9  ...  11  12  \\\n",
       "0  lnc_tr17504  Chr01  185390  186013  +  624  0  0  0  0  ...   0   1   \n",
       "1   lnc_tr4769  Chr01  201028  201445  +  418  0  0  0  0  ...   0   0   \n",
       "2   lnc_tr4790  Chr01  237391  237672  -  282  0  0  0  0  ...   0   0   \n",
       "3   lnc_tr4852  Chr01  627880  628299  +  420  0  0  0  0  ...   0   0   \n",
       "4  lnc_tr14183  Chr01  627880  628299  +  420  0  0  0  0  ...   0   0   \n",
       "\n",
       "                                               names g1_anti_rpm  g2_anti_rpm  \\\n",
       "0               >Pp3c1_350V3::Chr01:185389-186013(+)         0.0          0.0   \n",
       "1                         >lcl|Ppatens_Pp3c1_410V3.1         0.0          0.0   \n",
       "2                         >lcl|Ppatens_Pp3c1_500V3.1         0.0          0.0   \n",
       "3                         >lcl|Ppatens_Pp3c1_870V3.2         0.0          0.0   \n",
       "4  >PACX32968721 gene_name=Pp3c1_870 ref_gene_id=...         0.0          0.0   \n",
       "\n",
       "   g3_anti_rpm  g4_anti_rpm  p1_anti_rpm  p2_anti_rpm  p3_anti_rpm  \n",
       "0          0.0          0.0          0.0          0.0     0.552181  \n",
       "1          0.0          0.0          0.0          0.0     0.000000  \n",
       "2          0.0          0.0          0.0          0.0     0.000000  \n",
       "3          0.0          0.0          0.0          0.0     0.000000  \n",
       "4          0.0          0.0          0.0          0.0     0.000000  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anti[\"g1_anti_rpm\"] = anti.iloc[:,6]/0.393\n",
    "anti[\"g2_anti_rpm\"] = anti.iloc[:,7]/0.982\n",
    "anti[\"g3_anti_rpm\"] = anti.iloc[:,8]/0.77\n",
    "anti[\"g4_anti_rpm\"] = anti.iloc[:,9]/0.88\n",
    "anti[\"p1_anti_rpm\"] = anti.iloc[:,10]/1.597\n",
    "anti[\"p2_anti_rpm\"] = anti.iloc[:,11]/0.5656\n",
    "anti[\"p3_anti_rpm\"] = anti.iloc[:,12]/1.811\n",
    "anti.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>names</th>\n",
       "      <th>g1_anti_rpm</th>\n",
       "      <th>g2_anti_rpm</th>\n",
       "      <th>g3_anti_rpm</th>\n",
       "      <th>g4_anti_rpm</th>\n",
       "      <th>p1_anti_rpm</th>\n",
       "      <th>p2_anti_rpm</th>\n",
       "      <th>p3_anti_rpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_tr17504</td>\n",
       "      <td>&gt;Pp3c1_350V3::Chr01:185389-186013(+)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_tr4769</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_410V3.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_tr4790</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_500V3.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_tr4852</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_870V3.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_tr14183</td>\n",
       "      <td>&gt;PACX32968721 gene_name=Pp3c1_870 ref_gene_id=...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ids                                              names  \\\n",
       "0  lnc_tr17504               >Pp3c1_350V3::Chr01:185389-186013(+)   \n",
       "1   lnc_tr4769                         >lcl|Ppatens_Pp3c1_410V3.1   \n",
       "2   lnc_tr4790                         >lcl|Ppatens_Pp3c1_500V3.1   \n",
       "3   lnc_tr4852                         >lcl|Ppatens_Pp3c1_870V3.2   \n",
       "4  lnc_tr14183  >PACX32968721 gene_name=Pp3c1_870 ref_gene_id=...   \n",
       "\n",
       "   g1_anti_rpm  g2_anti_rpm  g3_anti_rpm  g4_anti_rpm  p1_anti_rpm  \\\n",
       "0          0.0          0.0          0.0          0.0          0.0   \n",
       "1          0.0          0.0          0.0          0.0          0.0   \n",
       "2          0.0          0.0          0.0          0.0          0.0   \n",
       "3          0.0          0.0          0.0          0.0          0.0   \n",
       "4          0.0          0.0          0.0          0.0          0.0   \n",
       "\n",
       "   p2_anti_rpm  p3_anti_rpm  \n",
       "0          0.0     0.552181  \n",
       "1          0.0     0.000000  \n",
       "2          0.0     0.000000  \n",
       "3          0.0     0.000000  \n",
       "4          0.0     0.000000  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anti = anti.drop(columns=[1,2,3,4,5,6,7,8,9,10,11,12])\n",
    "anti.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_rpm</th>\n",
       "      <th>g2_rpm</th>\n",
       "      <th>g3_rpm</th>\n",
       "      <th>g4_rpm</th>\n",
       "      <th>names</th>\n",
       "      <th>p1_rpm</th>\n",
       "      <th>p2_rpm</th>\n",
       "      <th>p3_rpm</th>\n",
       "      <th>g1_anti_rpm</th>\n",
       "      <th>g2_anti_rpm</th>\n",
       "      <th>g3_anti_rpm</th>\n",
       "      <th>g4_anti_rpm</th>\n",
       "      <th>p1_anti_rpm</th>\n",
       "      <th>p2_anti_rpm</th>\n",
       "      <th>p3_anti_rpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.298701</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.05499</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c1_40640V3.1</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>9.090909</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;XR_002970609.1 rna-XR_002970609.1_NCBI_ncRNA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&gt;lcl|Ppatens_Pp3c6_15880V3.1</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   g1_rpm   g2_rpm    g3_rpm  g4_rpm  \\\n",
       "0     0.0  0.00000  1.298701     0.0   \n",
       "1     0.0  3.05499  0.000000     0.0   \n",
       "2     0.0  0.00000  0.000000     0.0   \n",
       "3     0.0  0.00000  9.090909     0.0   \n",
       "4     0.0  0.00000  0.000000     0.0   \n",
       "\n",
       "                                           names    p1_rpm  p2_rpm    p3_rpm  \\\n",
       "0  >XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA  0.000000     0.0  0.000000   \n",
       "1  >XR_002972246.1 rna-XR_002972246.1_NCBI_ncRNA  0.626174     0.0  0.000000   \n",
       "2                   >lcl|Ppatens_Pp3c1_40640V3.1  0.626174     0.0  0.552181   \n",
       "3  >XR_002970609.1 rna-XR_002970609.1_NCBI_ncRNA  0.000000     0.0  0.000000   \n",
       "4                   >lcl|Ppatens_Pp3c6_15880V3.1  0.626174     0.0  0.552181   \n",
       "\n",
       "   g1_anti_rpm  g2_anti_rpm  g3_anti_rpm  g4_anti_rpm  p1_anti_rpm  \\\n",
       "0          0.0          0.0          0.0          0.0     0.000000   \n",
       "1          0.0          0.0          0.0          0.0     0.000000   \n",
       "2          0.0          0.0          0.0          0.0     0.626174   \n",
       "3          0.0          0.0          0.0          0.0     0.000000   \n",
       "4          0.0          0.0          0.0          0.0     0.000000   \n",
       "\n",
       "   p2_anti_rpm  p3_anti_rpm  \n",
       "0          0.0     0.000000  \n",
       "1          0.0     0.000000  \n",
       "2          0.0     0.552181  \n",
       "3          0.0     0.000000  \n",
       "4          0.0     0.000000  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now we merge all the tables\n",
    "\n",
    "features = outp.merge(anti, how=\"outer\", on=\"names\")\n",
    "features = features.drop(columns = [\"names_x\", \"names_y\", \"ids\"])\n",
    "features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XR_002967627.1</td>\n",
       "      <td>ref|XR_002967596.1|,ref|XR_002967627.1|::Chr22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|Ppatens_Pp3c2_1700V3.1</td>\n",
       "      <td>ref|XR_002973712.1|,lcl|Ppatens_Pp3c2_1700V3.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|Ppatens_Pp3c20_3270V3.2</td>\n",
       "      <td>lcl|Ppatens_Pp3c20_3270V3.18474,lcl|Ppatens_Pp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XR_002967369.1</td>\n",
       "      <td>ref|XR_002967365.1|,ref|XR_002967366.1|,ref|XR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|Ppatens_Pp3c13_20260V3.1</td>\n",
       "      <td>lcl|Ppatens_Pp3c13_20260V3.15922::Chr13:142488...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     transcript  \\\n",
       "0                XR_002967627.1   \n",
       "1    lcl|Ppatens_Pp3c2_1700V3.1   \n",
       "2   lcl|Ppatens_Pp3c20_3270V3.2   \n",
       "3                XR_002967369.1   \n",
       "4  lcl|Ppatens_Pp3c13_20260V3.1   \n",
       "\n",
       "                                               locus  \n",
       "0  ref|XR_002967596.1|,ref|XR_002967627.1|::Chr22...  \n",
       "1  ref|XR_002973712.1|,lcl|Ppatens_Pp3c2_1700V3.1...  \n",
       "2  lcl|Ppatens_Pp3c20_3270V3.18474,lcl|Ppatens_Pp...  \n",
       "3  ref|XR_002967365.1|,ref|XR_002967366.1|,ref|XR...  \n",
       "4  lcl|Ppatens_Pp3c13_20260V3.15922::Chr13:142488...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adding locuses of each transcript\n",
    "locuses = pd.read_csv('names_table_locus.csv', sep=\"\\t\")\n",
    "locuses.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_rpm</th>\n",
       "      <th>g2_rpm</th>\n",
       "      <th>g3_rpm</th>\n",
       "      <th>g4_rpm</th>\n",
       "      <th>transcript</th>\n",
       "      <th>p1_rpm</th>\n",
       "      <th>p2_rpm</th>\n",
       "      <th>p3_rpm</th>\n",
       "      <th>source</th>\n",
       "      <th>locus</th>\n",
       "      <th>g1_rpm_anti</th>\n",
       "      <th>g2_rpm_anti</th>\n",
       "      <th>g3_rpm_anti</th>\n",
       "      <th>g4_rpm_anti</th>\n",
       "      <th>p1_rpm_anti</th>\n",
       "      <th>p2_rpm_anti</th>\n",
       "      <th>p3_rpm_anti</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>lcl|Ppatens_Pp3c1_40640V3.1</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>flair</td>\n",
       "      <td>lcl|Ppatens_Pp3c1_40640V3.1580,PACX32971456::C...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.552181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>lcl|Ppatens_Pp3c6_15880V3.1</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>flair</td>\n",
       "      <td>ref|XR_002970176.1|,ref|XR_002970178.1|,ref|XR...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.544529</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.597403</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CNT2063702</td>\n",
       "      <td>2.504696</td>\n",
       "      <td>3.536068</td>\n",
       "      <td>0.552181</td>\n",
       "      <td>flair</td>\n",
       "      <td>lcl|Ppatens_Pp3c7_17330V3.23404,lcl|Ppatens_Pp...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CNT2063702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.536068</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>featureCounts</td>\n",
       "      <td>lcl|Ppatens_Pp3c7_17330V3.23404,lcl|Ppatens_Pp...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.896104</td>\n",
       "      <td>5.681818</td>\n",
       "      <td>lcl|Ppatens_Pp3c6_6770V3.4</td>\n",
       "      <td>6.887915</td>\n",
       "      <td>3.536068</td>\n",
       "      <td>5.521811</td>\n",
       "      <td>flair</td>\n",
       "      <td>lcl|Ppatens_Pp3c6_6770V3.42820,PACX32975550::C...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_rpm  g2_rpm    g3_rpm    g4_rpm                   transcript  \\\n",
       "0  0.000000     0.0  0.000000  0.000000  lcl|Ppatens_Pp3c1_40640V3.1   \n",
       "1  0.000000     0.0  0.000000  0.000000  lcl|Ppatens_Pp3c6_15880V3.1   \n",
       "2  2.544529     0.0  2.597403  0.000000                   CNT2063702   \n",
       "3  0.000000     0.0  0.000000  0.000000                   CNT2063702   \n",
       "4  0.000000     0.0  3.896104  5.681818   lcl|Ppatens_Pp3c6_6770V3.4   \n",
       "\n",
       "     p1_rpm    p2_rpm    p3_rpm         source  \\\n",
       "0  0.626174  0.000000  0.552181          flair   \n",
       "1  0.626174  0.000000  0.552181          flair   \n",
       "2  2.504696  3.536068  0.552181          flair   \n",
       "3  0.000000  3.536068  0.000000  featureCounts   \n",
       "4  6.887915  3.536068  5.521811          flair   \n",
       "\n",
       "                                               locus  g1_rpm_anti  \\\n",
       "0  lcl|Ppatens_Pp3c1_40640V3.1580,PACX32971456::C...          0.0   \n",
       "1  ref|XR_002970176.1|,ref|XR_002970178.1|,ref|XR...          0.0   \n",
       "2  lcl|Ppatens_Pp3c7_17330V3.23404,lcl|Ppatens_Pp...          0.0   \n",
       "3  lcl|Ppatens_Pp3c7_17330V3.23404,lcl|Ppatens_Pp...          0.0   \n",
       "4  lcl|Ppatens_Pp3c6_6770V3.42820,PACX32975550::C...          0.0   \n",
       "\n",
       "   g2_rpm_anti  g3_rpm_anti  g4_rpm_anti  p1_rpm_anti  p2_rpm_anti  \\\n",
       "0          0.0          0.0          0.0     0.626174          0.0   \n",
       "1          0.0          0.0          0.0     0.000000          0.0   \n",
       "2          0.0          0.0          0.0     0.000000          0.0   \n",
       "3          0.0          0.0          0.0     0.000000          0.0   \n",
       "4          0.0          0.0          0.0     0.000000          0.0   \n",
       "\n",
       "   p3_rpm_anti  \n",
       "0     0.552181  \n",
       "1     0.000000  \n",
       "2     0.000000  \n",
       "3     0.000000  \n",
       "4     0.000000  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features.to_csv(\"features.tsv\", index=False)\n",
    "!sed 's/>//' features.tsv > features_corrected.tsv\n",
    "features = pd.read_csv(\"features_corrected.tsv\")\n",
    "features = features.rename(columns={\"names\": \"transcript\"})\n",
    "features = features.merge(locuses, how=\"inner\", on=\"transcript\")\n",
    "features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Then just changing the columns order\n",
    "!awk '{print $5, $10, $1, $2, $3, $4, $6, $7, $8, $9, $11, $12, $13, $14, $15, $16, $17}' lnc_nanopore_all.tsv > lnc_nanopore_all_inorder.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fasta output\n",
    "\n",
    "Found isoforms were converted into fasta. Flair gives fasta as an output, we had only to rename the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pat = re.compile(r'lnc_tr\\d*')\n",
    "names = pd.read_csv('lnc_tr_names_tomynames.csv', sep=';', header=None)\n",
    "\n",
    "input = SeqIO.parse(\"lnc.isoforms.fa\", \"fasta\")\n",
    "\n",
    "with open(\"lnc_names.fasta\", \"w\") as outp:\n",
    "    for i in input:\n",
    "        for a in pat.findall(i.name):\n",
    "            print(names.loc[:, 0][names.loc[:, 1] == a])\n",
    "            i.id =  list(names.loc[:, 0][names.loc[:, 1] == a])[0]\n",
    "            i.description = \"mossss\"\n",
    "            print(i)\n",
    "            SeqIO.write(i, outp, \"fasta\")\n",
    "            \n",
    "!sed 's/mossss//' lnc_names.fasta > lnc_multiexon.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For monoexon features intersect was used.\n",
    "!bedtools merge -s -i guppy4all_sort.bed > guppy4all_sort_merged.bed \n",
    "!bedtools intersect  -s -loj  -a guppy4all_sort_merged.bed -b  lnc_annot_1ex.gff | grep gene_id > lnc_true_s.tsv \n",
    "!awk '{print $1, \"\\t\", $2,\"\\t\", $3,\"\\t\", $6,\"\\t\", $4}' lnc_true_s.bed | sed 's/ //g' > lnc_true_s_ord.bed\n",
    "!bedtools getfasta -name -fi nanomoss.fasta -bed lnc_true_s_ord.bed -fo lnc_monoex1.fasta\n",
    "#Then renaming as given above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stringtie\n",
    "\n",
    "Stringtie (https://github.com/gpertea/stringtie) was also used to prove lncRNA isoforms expression. We used both Flair and Stringtie results to compare with annotation.\n",
    "Gffcompare (https://github.com/gpertea/gffcompare) codes of .gtf.refmap output file indicate type of intersection with the annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!./stringtie -L -G lnc_annot.gtf -B -o stringtie_lnc.gtf lnc.bam # all reads aligned in one .bam file, produced by Flair\n",
    "\n",
    "!./gffcompare -R -r lnc_annot.gtf -o compare_lnc_stringtie stringtie_lnc.gtf\n",
    "!./gffcompare -R -r lnc_annot.gtf -o compare_lnc_flair lnc.gtf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
